### PR TITLE
Logger

### LINKED ISSUE #50: New logger subsystem
We need a better logging system.

Filtering messages, verbosity levels, origin tracking, backtraces, ...

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/5619887-new-logger-subsystem?utm_campaign=plugin&utm_content=tracker%2F371436&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F371436&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mic-e**
Finally, the new™ logging system, all ready to be reviewed to shreds.

Old `cpp/log.h` has been replaced by `cpp/log/log.h`, which provides the `MSG` macro and `openage::log::log()`.

Example usage:

    log::log(MSG(err) << "test error message " << 1337)
    my_unit.log(MSG(dbg) << "this message will be logged by the unit");
    throw util::Error(MSG(err) << "exceptions use the MSG system as well!")

See `doc/implementation/logger.md`.

Main step towards #48
Closes #50
Closes #171

### CODE REVIEW

File: cpp/crossplatform/timing.cpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+}
+#endif
+
+int64_t get_real_time() {
+	#ifdef __linux
+	// we don't need supreme accuracy; REALTIME_COARSE should suffice.
+	return posix_get_time(CLOCK_REALTIME_COARSE);
+	#elif __APPLE__
+	static_assert(false, "get_real_time not yet implemented for Apple");
+	#elif WIN32
```
Comment (TheJJ) [inline cpp/crossplatform/timing.cpp]: #209, `_WIN32` pls

File: cpp/crossplatform/timing.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+#endif
+
+namespace openage {
+namespace timing {
+
+#ifdef __linux
+int64_t posix_get_time(clockid_t clk_id) {
+	struct timespec t;
+	clock_gettime(clk_id, &t);
+	return ((long) t.tv_sec * 1000000000L) + (long) t.tv_nsec;
```
Comment (TheJJ) [inline cpp/crossplatform/timing.cpp]: c++casts

File: cpp/crossplatform/timing.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	static_assert(false, "unknown operating system");
+	#endif
+}
+
+int64_t get_monotonic_time() {
+	#ifdef __linux
+	return posix_get_time(CLOCK_MONOTONIC);
+	#elif __APPLE__
+	static_assert(false, "get_monotonic_time not yet implemented for Apple");
+	#elif WIN32
```
Comment (TheJJ) [inline cpp/crossplatform/timing.cpp]: `_WIN32`

File: cpp/crossplatform/timing.cpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	#else
+	static_assert(false, "unknown operating system");
+	#endif
+}
+
+int64_t get_monotonic_time() {
+	#ifdef __linux
+	return posix_get_time(CLOCK_MONOTONIC);
+	#elif __APPLE__
+	static_assert(false, "get_monotonic_time not yet implemented for Apple");
```
Comment (TheJJ) [inline cpp/crossplatform/timing.cpp]: until implemented otherwise, using the sdl api is probably a good workaround.

File: cpp/log/log.cpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+
+void log(MessageBuilder &msg) {
+	general_log_source->log(msg);
+}
+
+
+} // namespace log
+} // namespace openage
```
Comment (TheJJ) [inline cpp/log/log.cpp]: where's your new namespace style now?

File: cpp/log/message.cpp [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+	:
+	text{std::move(msg.text)},
+	meta{msg.meta} {}
+
+
+Message &Message::operator =(Message &&msg) {
+	this->text = std::move(msg.text);
+	this->meta = msg.meta;
+}
+*/
```
Comment (TheJJ) [inline cpp/log/message.cpp]: omgwtfnooblolgtfo commented out code commited!111

**Comment by franciscod** [comment]
niceeeeeee

**Pull request merged**